File: Fwd_Euler/driver_adapt_euler.cpp
/* Main routine to test adaptive forward Euler method on the scalar-valued ODE problem
     y' = -exp(-t)*y, t in [0,5],
     y(0) = 1.

   D.R. Reynolds
   Math 6321 @ SMU
   Fall 2020  */

#include <iostream>
#include <iomanip>
#include "adapt_euler.hpp"

using namespace std;
using namespace arma;


// ODE RHS function class -- instantiates a RHSFunction
class MyRHS: public RHSFunction {
public:
  int Evaluate(double t, vec& y, vec& f) {    // evaluates the RHS function, f(t,y)
    f = -exp(-t)*y;
    return 0;
  }
};
//    Convenience function for analytical solution
vec ytrue(const double t) {
  vec yt(1);
  yt(0) = exp(exp(-t)-1);
  return yt;
};


// main routine
int main() {

  // initial condition and time span
  vec y0("1.0");
  double t0 = 0.0;
  double Tf = 5.0;

  // set desired output times
  int Nout = 6;  // includes initial condition
  vec tspan = linspace(t0, Tf, Nout);

  // create ODE RHS function object
  MyRHS rhs;

  // create true solution object
  mat Ytrue(1,Nout);
  for (size_t i=0; i<Nout; i++)
    Ytrue.col(i) = ytrue(tspan(i));

  // tolerances
  vec rtols("1.e-3, 1.e-5, 1.e-7");
  double atol = 1.e-11;

  // create adaptive forward Euler stepper object (will reset rtol before each solve)
  AdaptEuler AE(rhs, 0.0, atol, y0);

  // loop over relative tolerances
  cout << "Adaptive Euler test problem, steps and errors vs tolerances:\n";
  for (size_t ir=0; ir<rtols.size(); ir++) {

    // update the relative tolerance, and call the solver
    cout << "  rtol = " << rtols(ir) << endl;
    AE.rtol = rtols(ir);
    mat Y = AE.Evolve(tspan, y0);

    // output solution, errors, and overall error
    mat Yerr = abs(Y-Ytrue);
    for (size_t i=0; i<Nout; i++)
      printf("    y(%.1f) = %9.6f,   abserr = %.2e,  relerr = %.2e\n",
             tspan(i), Y(0,i), Yerr(0,i), Yerr(0,i)/abs(Y(0,i)));
    cout << "  overall: "
         << "\t steps = " << AE.steps
	       << "\t fails = " << AE.fails
	       << "\t abserr = " << norm(Yerr,"inf") 
	       << "\t relerr = " << norm(Yerr/Ytrue,"inf") 
	       << endl;

  }

  return 0;
}

File: forward_euler/driver_adapt_euler.py
#!/usr/bin/env python3
#
# Main routine to test adaptive forward Euler method on the scalar-valued ODE problem
#    y' = -exp(-t)*y, t in [0,5],
#    y(0) = 1.
#
# D.R. Reynolds
# Math 6321 @ SMU
# Fall 2023

import numpy as np
from AdaptEuler import *

# problem time interval
t0 = 0.0
tf = 5.0

# problem-definining functions
def f(t,y):
    """ ODE RHS function """
    return np.array([-np.exp(-t)*y[0]])
def ytrue(t):
    """ Analytical solution """
    return np.array([np.exp(np.exp(-t)-1.0)])

# shared testing data
Nout = 6   # includes initial condition
tspan = np.linspace(t0, tf, Nout)

# get true solution at output times
Ytrue = np.zeros((Nout,1))
for i in range(Nout):
    Ytrue[i,:] = ytrue(tspan[i])
y0 = Ytrue[0,:]

# set the testing tolerances
rtols = np.array((1.e-3, 1.e-5, 1.e-7))
atol = 1.e-11

# create adaptive forward Euler stepper object (will reset rtol before each solve)
AE = AdaptEuler(f, y0)

# loop over relative tolerances
print("Adaptive Euler test problem, steps and errors vs tolerances:")
for rtol in rtols:

    # set the relative tolerance, and call the solver
    print("  rtol = ", rtol)
    AE.set_rtol(rtol)
    Y, success = AE.Evolve(tspan, y0)
    if (not success):
        print("    solve failed at this tolerance")
        continue

    # output solution, errors, and overall error
    Yerr = np.abs(Y-Ytrue)
    for i in range(Nout):
        print("    y(%.1f) = %9.6f,   abserr = %.2e,  relerr = %.2e" %
              (tspan[i], Y[i,0], Yerr[i,0], Yerr[i,0]/np.abs(Y[i,0])))
    print("  overall:  steps = %5i  fails = %2i  abserr = %9.2e  relerr = %9.2e\n" %
          (AE.get_num_steps(), AE.get_num_error_failures(), np.linalg.norm(Yerr,np.inf),
           np.linalg.norm(Yerr/Ytrue,np.inf)))
